src/cargo/core/compiler/mod.rs

```diff
@@ -656,7 +637,14 @@ fn rustdoc<'a, 'cfg>(cx: &mut Context<'a, 'cfg>, unit: &Unit<'a>) -> CargoResult
         }
         state.running(&rustdoc);
 
-        let exec_result = if should_capture_output {
+        let exec_result = if json_messages {
+            rustdoc
+                .exec_with_streaming(
+                    &mut json_stdout,
+                    &mut |line| json_stderr(line, &package_id, &target),
+                    false,
+                ).map(drop)
```

Comment:
`map(drop)` probably should not be here, it's just ignore the error. I wonder if the one on the next line should be removed as well?

Reply:
Yea, it's a little weird.  The intent is so that all branches are `CargoResult<()>` so they all have the same type so that a common error can be attached.  I can't think of a more elegant way to do that.

Reply:
Oh, I've misunderstood the intention of the code completely... 

Yep, the only better way to write this would probably be a `catch` block, which is currently unstable.

Reply:
You could go the other way and change `rustdoc.exec()` to `rustdoc.exec_with_output()`, so all branches type-check to `CargoResult<Output>`.

Reply:
That'll unfortunately hide rustdoc's output from the user, b/c it would be captured.

Reply:
Ah. I see.